[torch.func] Add docs

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91319

Docs copy-pasted from functorch docs with minor adjustments. We are
keeping the functorch docs for BC, though that's up for debate -- we
could also just say "see .. in torch.func" for some, but not all doc
pages (we still want to keep around any examples that use
make_functional so that users can tell what the difference between that
and the new functional_call is).

Test Plan:
- docs preview

zou3519 committed:
[torch.func] Add docs

Docs copy-pasted from functorch docs with minor adjustments. We are
keeping the functorch docs for BC, though that's up for debate -- we
could also just say "see .. in torch.func" for some, but not all doc
pages (we still want to keep around any examples that use
make_functional so that users can tell what the difference between that
and the new functional_call is).

Test Plan:
- docs preview

[ghstack-poisoned]

zou3519 committed:
Update on "[torch.func] Add docs"


Docs copy-pasted from functorch docs with minor adjustments. We are
keeping the functorch docs for BC, though that's up for debate -- we
could also just say "see .. in torch.func" for some, but not all doc
pages (we still want to keep around any examples that use
make_functional so that users can tell what the difference between that
and the new functional_call is).

Test Plan:
- docs preview

[ghstack-poisoned]

zou3519 committed:
Update on "[torch.func] Add docs"


Docs copy-pasted from functorch docs with minor adjustments. We are
keeping the functorch docs for BC, though that's up for debate -- we
could also just say "see .. in torch.func" for some, but not all doc
pages (we still want to keep around any examples that use
make_functional so that users can tell what the difference between that
and the new functional_call is).

Test Plan:
- docs preview

[ghstack-poisoned]

samdow commented on docs/source/func.whirlwind_tour.rst:
```diff
@@ -0,0 +1,197 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Note: :func:`vmap` imposes restrictions on the code that it can be used on. For more
+details, please read its docstring.
```
nit: Since it's spelled out in more detail than in the docstring, we should link to the UX limitations here

samdow commented on docs/source/func.whirlwind_tour.rst:
```diff
@@ -0,0 +1,197 @@
+    x = torch.randn(5)
+    y = torch.randn(5)
+    f = lambda x, y: (x * y)
+    _, output = jvp(f, (x, y), (torch.ones(5), torch.ones(5)))
```
nit: since above `outputs` is the result of the forward computation, it's a little confusing to have it be the result of the forward AD computation here.

Maybe we call them `primals` and `tangents` [like JAX](https://jax.readthedocs.io/en/latest/_autosummary/jax.jvp.html?highlight=jvp) (i.e. `primals, tangents = jvp(...)`

samdow commented on docs/source/func.whirlwind_tour.rst:
```diff
@@ -0,0 +1,197 @@
+    expected = torch.diag(torch.cos(x))
+    assert torch.allclose(jacobian, expected)
+
+Use :func:`jacrev` to compute the jacobian. This can be composed with vmap to produce
```
Nit: the wording is a little weird here. Maybe it meant to be either: "`jacrev` can be composed with ..."

samdow commented on docs/source/func.whirlwind_tour.rst:
```diff
@@ -0,0 +1,197 @@
+      return x.sin().sum()
+
+      x = torch.randn(5)
+      hess = hessian(f)(x)
```
nit: indentation-- right now it's

```
def f(x):
  return x.sin().sum()
  x = torch.randn(5)
  hess = hessian(f)(x)
```
and it should be
```
def f(x):
  return x.sin().sum()
  x = torch.randn(5)
hess = hessian(f)(x)
```

samdow commented on docs/source/func.rst:
```diff
@@ -1,7 +1,54 @@
+- A "function transform" is a higher-order function that accepts a numerical function
+  and returns a new function that computes a different quantity.
+
+- functorch has auto-differentiation transforms (``grad(f)`` returns a function that
```
nit: torch.func

samdow commented on docs/source/func.ux_limitations.rst:
```diff
@@ -0,0 +1,293 @@
+will print "hello!" once and pop only one element from ``lst``.
+
+
+:func:`vmap` executes `f` a single time, so all side effects only happen once.
```
nit: 
```
`f` => ``f`` 
```

samdow commented on docs/source/func.ux_limitations.rst:
```diff
@@ -0,0 +1,293 @@
+  x = torch.randn(1)
+  y = torch.randn(3)
+
+  # Raises an error because `y` has fewer elements than `x`.
```
I think this is backwards and I think someone brought this up before and we just completely forgot to change it 🫢

Also IIRC someone said something that this is confusing since people see it add `torch.randn(1).add_(torch.randn(3))` which never works instead of a for loop over `torch.randn(1).add_(torch.randn(()))` which is what it's trying to show doesn't work. I'm not sure but it might be easier to under if we had 
`y = torch.randn(3, 1)  # when vmaped over, looks to have size of [1]`

zou3519 replied:
Ah yes you're right, thanks for pointing it out!

zou3519 replied:
I've been meaning to talk about zeros -> new_zeros since I tell people about that a lot, so, will also drop a section in here about that

samdow commented on docs/source/func.ux_limitations.rst:
```diff
@@ -0,0 +1,293 @@
+three elements back into ``x``, which only has one element, raises an error
+due to attempting to write three elements into a Tensor with a single element.
+
+There is no problem if the Tensor being written to has the same number of
+elements (or more):
```
nit: I think this might mean to say "if the Tensor being written into is batched under vmap" since technically it doesn't work to have

```
def f(x, y):
  x.add_(y)
  return x

x = torch.randn(2, 3)
y = torch.randn(2, 3)

vmap(f, in_dims=(None, 0)(x, y)
```

samdow reviewed (approved):
Looks good! I know this is just puling it over so if you want to leave the comments as follow up, that's fine with me. Just things I noticed since we're in here

zou3519 commented:
> Looks good! I know this is just puling it over so if you want to leave the comments as follow up, that's fine with me. Just things I noticed since we're in here

Thanks! I'll address your comments and improve things in this PR while I'm at it

zou3519 committed:
Update on "[torch.func] Add docs"


Docs copy-pasted from functorch docs with minor adjustments. We are
keeping the functorch docs for BC, though that's up for debate -- we
could also just say "see .. in torch.func" for some, but not all doc
pages (we still want to keep around any examples that use
make_functional so that users can tell what the difference between that
and the new functional_call is).

Test Plan:
- docs preview

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
